### PR TITLE
Make e-mail sanity check more precise

### PR DESCRIPTION
**Pull request by balanceofcowards**
For security reasons, all e-mail addresses starting with non-alphanumeric characters were rejected. This is too broad and rejects perfectly valid e-mail addresses. Only leading hyphens should be rejected -- in all other cases e-mail address specification should follow RFC 5322.

### CODE REVIEW

**Review by delvh — APPROVED** [review]

**Comment by balanceofcowards** [comment]
... overlooked the unit tests, sorry.
Having a look at the tests: If e-mail address strings are indeed forwarded to the shell literally, then the dollar (`$`) and backtick (`` ` ``) characters should be rejected everywhere in the address, as they could be interpreted in the shell (cf. `$HOME@example.com`, `` `pwd`@example.com``).

It would be far better to escape user provided strings properly, before handing them to the shell, though. I don't know enough about the exact problem here, though. Please advise.

**Comment by silverwind** [comment]
I think we really just should just use [`mail.ParseAddress`](https://stackoverflow.com/a/66624104/808699). See https://github.com/go-gitea/gitea/pull/17688 for more context.

**Comment by lunny** [comment]
> I think we really just should just use [`mail.ParseAddress`](https://stackoverflow.com/a/66624104/808699), everything else is madness. See #17688 for more context.

We can't do that. That's what we changed from and we change it because of some security report about that method.

**Comment by silverwind** [comment]
> We can't do that. That's what we changed from and we change it because of some security report about that method.

Have these "security issues" been raised to golang? Can you provide examples of what was the problem?

**Comment by lunny** [comment]
> > We can't do that. That's what we changed from and we change it because of some security report about that method.
> 
> Have these "security issues" been raised to golang? Can you provide examples of what was the problem?

See #19102 and #17688

**Comment by silverwind** [comment]
We ought to just correctly pass `--` to sendmail in all cases to fix that. We should not have to restrict emails just because we are using tools like sendmail incorrectly.

**Comment by silverwind** [comment]
Generally I think we should remove `sendmail` integration and just do SMTP directly in golang. `sendmail` is a antiquated tool and there are many different implementations that may not be compatible with each other, it's just a pain in the ass 😉.

**Comment by balanceofcowards** [comment]
From an external viewpoint, I totally agree with @silverwind. Meanwhile, as a stopgap solution, the sendmail problem can be remediated by checking only for a leading hyphen character.

Regarding my previous remarks about the `$` and `` ` `` characters: I realized this would only pose a problem if the command was called in a shell context (which is a bad idea, due to exactly this problem). Without knowing details about your code, sendmail likely is not called this way: I tried a basic smoketest on the live test instance and, e.g. `a$HOME@domain` was processed correctly.

Unfortunately, with the restrictions in place, Gitea is totally broken for me as a user. I cannot access my account and I have no sane way to remediate the problem. I probably could scan the database and replace the offending e-mail address manually... but I would rather not go down that road, fearing what else might break along the way.

**Comment by silverwind** [comment]
Indeed, we should never run commands in shell context as it's pretty much impossible to escape arguments safely for a shell. We should always use `exec.Command(cmd, ...args)` format which is safe as long as `cmd` is not a shell.

**Comment by silverwind** [comment]
So regarding these changes, they seem okay. But there are various characters that have special meanings in shell context. Can it be demonstrated that this is safe to use with `sendmail`? Long them, we should of course move away from sendmail to SMTP only.

**Comment by balanceofcowards** [comment]
> So regarding these changes, they seem okay. But there are various characters that have special meanings in shell context. Can it be demonstrated that this is safe to use with `sendmail`? [...]

In principle, that leads to the same problem @zeripath already outlined in #19102: To be sure, one would have to test every sendmail implementation out there. However, I would consider it unlikely that any sendmail argparser processes special characters other than hyphens.

Regarding characters interpreted by the *shell* (not sendmail), at first a similar problem would arise: There are different shells out there, which likely have different special characters. The obviously "dangerous" characters would be:

* The dollar sign `$` (as `$HOME@domain` might reveal the home dir of the current process);
* The exclamation mark `!` (as `!13@domain` might reveal command nr. 13 in the history list); 
* And the back tick `` ` `` (as `` `pwd`@domain `` might execute any command).

Note, that these characters do not necessarily depend on being in the first position: `$HOME@domain` would be as vulnerable as `a$HOME@domain`. All of these, however, are easily circumvented by not executing external commands in a shell context, in the first place (and I don't think you're doing that, anyway). The second best option would be to escape the characters properly.

In summary, my argument for the PR is: From a security perspective, it does not decrease the status quo, but it fixes the erroneous rejection of valid e-mail addresses for a large number of potential addresses.

**Comment by silverwind** [comment]
I see we call sendmail with

```go
exec.CommandContext(ctx, setting.MailService.SendmailPath, args...)
```

Which is save from shell interpretion at least. And anything beyond that are sendmail bugs, so LGTM.

**Review by silverwind — APPROVED** [review]

File: models/user/email_address.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ func (err ErrEmailCharIsNotSupported) Error() string {
 }
 
 // ErrEmailInvalid represents an error where the email address does not comply with RFC 5322
+// or has a leading '-' character
```
Comment (techknowlogick) [inline models/user/email_address.go:44]: I'm not sure - should be the only forbidden leading character, because the issue is people using sendmail to send their email, and the interface with it is prone to turning arguments into RCEs

**Review by techknowlogick — DISMISSED** [review]

File: models/user/email_address.go (hunk 1 above)
Comment (techknowlogick) [reply models/user/email_address.go:44]: update, or just apply the strict filtering if the sendmail interface is used?

Comment (balanceofcowards) [reply models/user/email_address.go:44]: Assuming that sendmail implementations are a black box, potentially any character in an e-mail address could be dangerous. Realistically, though, only the dash character will generate problems. I am not aware of any POSIX-style command line parser that treats any character beyond '-' as special (well, except for 2nd-level 'key:value' mappings such as seen in ffmpeg or mplayer... but that doesn't really apply here).

Performing the check only when sendmail is used might work, but will likely complicate the code. I would suggest that in that case @silverwind is correct, and sendmail should be scratched in favor of native SMTP (getting rid of the check, altogether).

Comment (silverwind) [reply models/user/email_address.go:44]: Yes, we should definitely evaluate dropping sendmail support if the arguments can not be safely escaped.

Comment (zeripath) [reply models/user/email_address.go:44]: I suspect we should just have a few options:

1) Allow everything but initial `-`
2) Only allow safe characters

This can be an option on the sendmail configuration and then the problem is no longer ours to solve and it becomes the system admins job to deal with it.

Comment (delvh) [reply models/user/email_address.go:44]: I'm not quite sure what you mean with `2.`.
I'd change `2.` a bit (`1.` is fine):
`2. Allow every valid email, leave it up to the admin to allow sendmail configurations`
`3. Allow every valid email, drop sendmail support`

**Comment by techknowlogick** [comment]
ping LG-TM

**Comment by cortices** [comment]
This is currently a blocking issue in accessing my gitea install due to the incorrect email validation on my admin account. Can this be brought into a release version soon please?

**Comment by wxiaoguang** [comment]
If it's urgent, to bypass, you can update the admin's e-mail in the database.
